src/read/magic_finder.rs

```diff
@@ -0,0 +1,191 @@
+use std::io::{Read, Seek, SeekFrom};
+
+use memchr::memmem::FinderRev;
+
+use crate::result::ZipResult;
+
+/// A utility for finding magic symbols from the end of a seekable reader.
+///
+/// Can be repurposed to recycle the internal buffer.
+pub struct MagicFinder<'a> {
+    buffer: Box<[u8]>,
+    pub(self) finder: FinderRev<'a>,
+    cursor: u64,
+    mid_buffer_offset: Option<usize>,
+    bounds: (u64, u64),
+}
+
+impl<'a> MagicFinder<'a> {
+    /// Create a new magic bytes finder to look within specific bounds.
+    pub fn new(magic_bytes: &'a [u8], bounds: (u64, u64)) -> Self {
```

Comment:
Split this into two parameters, `start_inclusive` and `end_exclusive`, for readability.